### PR TITLE
Adds better type narrowing docs

### PR DESCRIPTION
**Pull request by sobolevn**
List of changed:
- I've renamed `casts` into `type-narrowing`
- I've moved `TypeGuard` type there, it fits perfectly
- I've added new docs about general type narrowing techniques
- I've added / changed links where it was required

Refs https://github.com/python/mypy/pull/11014
CC @hauntsaninja

### CODE REVIEW

File: docs/source/type_narrowing.rst [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+Type narrowing
+==============
+
+This section is dedicated to  several type narrowing
+techniques which are supported by Mypy.
+
+What is type narrowing?
+It is an operation of telling type checking that some
+broader type is actually more specific.
```
Comment (hauntsaninja) [inline docs/source/type_narrowing.rst:11]:
```suggestion
Type narrowing is when you convince a type checker that a broader type is actually more specific, for instance, that an object of type ``Shape`` is actually of the narrower type ``Square``.
```

File: docs/source/type_narrowing.rst [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+       t = type(o)  # We must use a variable here
+       reveal_type(t)  # Revealed type is "builtins.type"
+
+       if issubtype(t, MyCalcMeta):  # `issubtype(type(o), MyCalcMeta)` won't work
+           reveal_type(t)  # Revealed type is "Type[MyCalcMeta]"
+           t.calc()  # Okay
+
+.. note::
+
+  Narrowing type for some impossible state will be treated as an error.
```
Comment (hauntsaninja) [inline docs/source/type_narrowing.rst:82]: This isn't generally true, e.g. not even with `--strict`, I think you need `--warn-unreachable` for this.

Comment (hauntsaninja) [reply docs/source/type_narrowing.rst:82]: Could also maybe mention the `assert_never` trick here though

File: docs/source/type_narrowing.rst [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+
+Type narrowing expressions
+--------------------------
+
+The simplest way to narrow a type is to use one of the supported expressions:
+
+- :py:func:`type` like in ``type(obj) is int`` will narrow ``obj`` to have ``int`` type
+- :py:func:`isinstance` like in ``isinstance(obj, float)`` will narrow ``obj`` to have ``float`` type
+- :py:func:`issubclass` like in ``issubclass(cls, MyClass)`` will narrow ``cls`` to be ``Type[MyClass]``
```
Comment (hauntsaninja) [inline docs/source/type_narrowing.rst:21]:
```suggestion
- :py:func:`isinstance` like in ``isinstance(obj, float)`` will narrow ``obj`` to have ``float`` type
- :py:func:`issubclass` like in ``issubclass(cls, MyClass)`` will narrow ``cls`` to be ``Type[MyClass]``
- :py:func:`type` like in ``type(obj) is int`` will narrow ``obj`` to have ``int`` type
```

File: docs/source/type_narrowing.rst [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+It is important to understand that type narrowing happens in a specific context.
+It can be a single ``if`` branch:
```
Comment (hauntsaninja) [inline docs/source/type_narrowing.rst:24]:
```suggestion
Type narrowing is contextual. For example, based on the condition, mypy will narrow an expression only within an ``if`` branch:
```

File: docs/source/type_narrowing.rst [hunk 5]
```diff
@@ -0,0 +1,289 @@
+.. _type-narrowing:
+
+Type narrowing
+==============
+
+This section is dedicated to  several type narrowing
+techniques which are supported by Mypy.
```
Comment (hauntsaninja) [inline docs/source/type_narrowing.rst:7]:
```suggestion
techniques which are supported by mypy.
```

File: docs/source/type_narrowing.rst [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+- :py:func:`issubclass` like in ``issubclass(cls, MyClass)`` will narrow ``cls`` to be ``Type[MyClass]``
+
+It is important to understand that type narrowing happens in a specific context.
+It can be a single ``if`` branch:
+
+.. code-block:: python
+
+  def function(arg: object):
+      if isinstance(arg, int):
+          # Type is narrowed for this ``if`` branch context only
```
Comment (hauntsaninja) [inline docs/source/type_narrowing.rst:30]:
```suggestion
          # Type is narrowed within the ``if`` branch only
```

File: docs/source/type_narrowing.rst [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+          reveal_type(arg)  # Revealed type: "builtins.int"
+      elif isinstance(arg, str) or isinstance(arg, bool):
+          # This ``if`` branch has new context for narrowing:
```
Comment (hauntsaninja) [inline docs/source/type_narrowing.rst:33]:
```suggestion
          # Type is narrowed differently within this ``elif`` branch:
```

File: docs/source/type_narrowing.rst [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+          reveal_type(arg)  # Revealed type: "builtins.str | builtins.bool"
+
+          # Contexts can be nested:
```
Comment (hauntsaninja) [inline docs/source/type_narrowing.rst:36]:
```suggestion
          # Subsequent narrowing operations will narrow the type further
```

File: docs/source/type_narrowing.rst [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+          if isinstance(arg, bool):
+              reveal_type(arg)  # Revealed type: "builtins.bool"
+
+      # In the main context type won't be changed:
```
Comment (hauntsaninja) [inline docs/source/type_narrowing.rst:40]:
```suggestion
      # Back outside of the ``if`` statement, the type isn't narrowed:
```

File: docs/source/type_narrowing.rst [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+      reveal_type(arg)  # Revealed type: "builtins.object"
+
+We can also use reverse logic as well:
```
Comment (hauntsaninja) [inline docs/source/type_narrowing.rst:43]:
```suggestion
Mypy understands the implications `return` or exception raising can have for what type an object could be:
```

**Review by hauntsaninja — COMMENTED** [review]
This is awesome. I made a bunch of small wording changes. The only substantive comment here is about `--warn-unreachable` which we need to be a little careful about. E.g., there's some discussion of it here: https://mypy.readthedocs.io/en/stable/common_issues.html#unreachable-code

**Comment by sobolevn** [comment]
@hauntsaninja thanks a lot for your help! I've added a note about `--warn-unreachable`

**Comment by hauntsaninja** [comment]
Thank you!

**Pull request merged**